imsim/photon_pooling.py

```diff
@@ -0,0 +1,350 @@
+    """
+    shape = bounds.numpyShape()
+    if shape[1] % 2 == 0:
+        dx = 0.5
+    else:
+        dx = 0.0
+    if shape[0] % 2 == 0:
+        dy = 0.5
+    else:
+        dy = 0.0
+    adjustment = galsim._PositionD(dx,dy)
+    return adjustment
+
+def offset_photon_arrays(stamps, batch, offset_adjustment):
+    """Take a batch of photon arrays and perform the sub-pixel offset needed to place
+    the object correctly within the full image.
+
+    Parameters:
+        stamps: a list of photon arrays generated by the objects in a batch
+        batch: a list of StellarObjects making up a batch
+        offset_adjustment: a galsim.PositionD by which to adjust each object offset
+    """
+    for photons, object in zip(stamps, batch):
+        offset = offset_adjustment - object.offset
+        photons.x += offset.x
+        photons.y += offset.y
+
+def accumulate_photons(photons, image, sensor, center):
+    """Accumulate a photon array onto a sensor.
+
+    Parameters:
+        photons: A PhotonArray containing the photons to be accumulated.
+        image: The image to which we draw the accumulated photons.
+        sensor: Sensor to use for accumulation. If None, a temporary sensor is created here.
+        center: Center of the image as galsim.PositionI.        
+    """
+    if sensor is None:
+        sensor = Sensor()
+    imview = image._view()
+    imview._shift(-center)  # equiv. to setCenter(), but faster
+    imview.wcs = PixelScale(1.0)
+    if imview.dtype in (np.float32, np.float64):
+        sensor.accumulate(photons, imview, imview.center)
+    else:
+        # Need a temporary
+        im1 = galsim.image.ImageD(bounds=imview.bounds)
+        sensor.accumulate(photons, im1, imview.center)
+        imview += im1
+
+
+def make_batches(objects, nbatch: int):
+    """Generator converting an input list of objects to batches.
+
+    Parameters:
+        objects: List of objects to be yielded in batches.
+        nbatch: The number of batches to create.
+
+    Yields:
+        A single batch made up of a list of object numbers.
+    """
+    per_batch = len(objects) // nbatch
+    o_iter = iter(objects)
+    for _ in range(nbatch):
+        yield [obj for _, obj in zip(range(per_batch), o_iter)]
+
+
+def build_stamps(base, logger, objects: list[StellarObject], stamp_type: str):
+    """Create stamps for a list of StellarObjects.
+
+    Parameters:
+        base: The base configuration dictionary.
+        logger: Logger object.
+        objects: List of StellarObjects for which we will build stamps.
+        stamp_type: The stamp type to be used.
+
+    Returns:
+        images: Tuple of the output stamps. In normal PhotonPooling usage these will actually be
+            PhotonArrays to be processed into images after they have been pooled following this step.
+        current_vars: Tuple of variables for each stamp (noise etc).
+    """
+    base["stamp"]["type"] = stamp_type
```

Comment:
What's happening here is that LSST_PhotonPoolingImage uses two stamp types: this function is called first to draw the FFT objects with `stamp_type='LSST_Silicon'`, and then for the photon pooling with `stamp_type='PhotonStampBuilder'` (to be renamed LSST_Photons) in which case it generates the photons but doesn't draw the image. Of course this also means that the stamp type in the config YAML is actually ignored, which perhaps isn't great, either. I wonder whether it might be possible to have a new type of stamp which does both the FFT drawing and the photon generation depending.

Reply:
Yeah, don't ignore the config stamp type.  PhotonStampBuilder is your new stamp type, so it needs to be able to handle both types of things.

Reply:
Done -- a new stamp type `LSST_Photons` should be used and handle all objects when using photon pooling and draws/just returns photons accordingly.